src/XGrid.cu

```diff
@@ -0,0 +1,260 @@
+/* 
+
+Implements the grid-multiplication onto a GPU using CUDA. 
+
+*/
+#include <iostream>
+#include "bifrost/XGrid.h"
+#include "assert.hpp"
+#include "trace.hpp"
+#include "utils.hpp"
+#include "cuda.hpp"
+#include "cuda/stream.hpp"
+#include "Complex.hpp"
+
+struct __attribute__((aligned(1))) nibble2 {
+    // Yikes!  This is dicey since the packing order is implementation dependent!  
+    signed char y:4, x:4;
+};
+
+struct __attribute__((aligned(1))) blenib2 {
+    // Yikes!  This is dicey since the packing order is implementation dependent!
+    signed char x:4, y:4;
+};
+
+template<typename In, typename Out>
+__global__ void XGrid_kernel(int npol, int gridsize, int nbatch, int nchan,
+		     const In* __restrict__  d_in,
+                     Out* d_out){
+
+        int bid_x = blockIdx.x, bid_y = blockIdx.y, bid_z = blockIdx.z ;
+        int blk_x = blockDim.x ;
+        int grid_x = gridDim.x, grid_y = gridDim.y , grid_z = gridDim.z ;
+        int tid_x = threadIdx.x ;
+	int pol_skip = grid_z*blk_x;
+    
+        extern  __shared__ Complex<float> shared[] ;
+        In* xx = reinterpret_cast<In *>(shared);
+        In* yy = xx + blk_x;
+ 
+        int tt = 1;
+        if(npol>1) tt=(int)npol/2;
+
+        int bid  = ((bid_x * grid_y + bid_y) * tt  * grid_z  + bid_z) * blk_x ;
+	int bid2 = ((bid_x * grid_y + bid_y) * npol * grid_z  + bid_z) * blk_x ;
+
+        #pragma unroll
+	for(int i=0;i<npol;i++){
+
+                xx[tid_x] = d_in[bid+i/2*pol_skip+tid_x];
+		yy[tid_x] = d_in[bid+i%2*pol_skip+tid_x];
+	        
+		d_out[bid2+i*pol_skip+tid_x].x += xx[tid_x].x*yy[tid_x].x + xx[tid_x].y*yy[tid_x].y;  
+	       	d_out[bid2+i*pol_skip+tid_x].y += xx[tid_x].y*yy[tid_x].x - xx[tid_x].x*yy[tid_x].y;
```

Comment:
Just want to double check... this function appears to be doing `x * conj(y)`. I only want to clarify because the `Complexfcma` function in `VGrid.cu` does `conj(x) * y`. Are these both correct, and what we want to be doing?

Reply:
`Complexfcma` is used to grid voltage data, and it's doing conj(kernel) * voltage. Here we are forming cross-pols of the entire grid (XX, XY, YX, YY), where it's convention that the second is conjugated. So this is correct.